Add decimals to int when using `into string --decimals`

x3rAx opened the pull request:
# Description

This is a follow up to PR https://github.com/nushell/nushell/pull/6084. It was intended as a fix to https://github.com/nushell/nushell/issues/6078 but @merelymyself beat me to it.

However, this PR adds the decimal digits in a way that they get also added when `group_digits` is set to `true` (see: https://github.com/nushell/nushell/issues/6078#issuecomment-1189636421).

I also added tests for `into string` when converting int with `--decimals`.

# Tests

Make sure you've done the following:

- [x] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [x] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [ ] ~If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.~

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [x] `cargo test --workspace --features=extra` to check that all the tests pass

x3rAx commented:
I can't execute the `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` command because it fails with an error:

```
error: cannot borrow `*grid` as mutable because it is also borrowed as immutable
   --> crates/nu-table/src/table.rs:302:13
    |
301 |         if let Some(line) = grid.get_split_line(1) {
    |                             ---------------------- immutable borrow occurs here
302 |             grid.set_split_line(grid.count_rows() - 1, line.clone());
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^------------^
    |             |                                          |
    |             |                                          immutable borrow later used here
    |             mutable borrow occurs here
    |
    = note: `-D mutable-borrow-reservation-conflict` implied by `-D warnings`
    = warning: this borrowing pattern was not meant to be accepted, and may become a hard error in the future
    = note: for more information, see issue #59159 <https://github.com/rust-lang/rust/issues/59159>

error: could not compile `nu-table` due to previous error
```

WindSoilder commented on crates/nu-command/src/conversions/into/string.rs:
```diff
@@ -305,21 +292,31 @@ pub fn action(
-fn format_int(int: i64) -> String {
-    int.to_string()
+fn format_int(int: i64, group_digits: bool, decimals: usize) -> String {
+    let str = if group_digits {
```
It seems that `group_digits` doesn't do any useful thing, this is why clippy throw out warning

WindSoilder commented on crates/nu-command/tests/commands/into_string.rs:
```diff
@@ -0,0 +1,49 @@
+use nu_test_support::{nu, pipeline};
```
We don't really need this file, these testing code can be moved to `crates/nu-command/tests/commands/str_/into_string.rs`

WindSoilder commented on crates/nu-command/tests/commands/mod.rs:
```diff
@@ -30,6 +30,7 @@ mod histogram;
 mod insert;
 mod into_filesize;
 mod into_int;
+mod into_string;
```
So this module is not really needed

fdncred commented:
Isn't think supposed to support a locale's digit delimiter? I believe I did the same thing with `format_filesize()` using https://github.com/bcmyers/num-format. You might want to look at that code for help. That could perhaps be a utility function in nu-utils as long as it doesn't require any nu values.

x3rAx commented:
I copied the part of `format_filesize()` that gets the system locale:

https://github.com/nushell/nushell/blob/8030f7e9f003bc90c67a2f3d386ec1ff4d7e1a25/crates/nu-protocol/src/value/mod.rs#L2352-L2370

It's working also with `group_digits` set to `true` but now I have two issues:

1. Where is the best place to move that code so that it is not duplicated?
2. How can I make sure the tests work independently of the system locale?

   When the system locale is set to `de_DE`, the decimal separator is a `,` (comma). When running the tests with `LC_ALL` set to `de_DE`, the tests fail because they expect a `.` (period) as decimal separator.

3. How can I write a test that ensures that the system locale is respected?

fdncred commented:
Thanks for working on this!

> Where is the best place to move that code so that it is not duplicated?

I'd put it in nu-utils as long as it has no nushell types like Value

> How can I make sure the tests work independently of the system locale?
> When the system locale is set to de_DE, the decimal separator is a , (comma). When running the tests with LC_ALL set to de_DE, the tests fail because they expect a . (period) as decimal separator.

I'm pretty sure the decimal separator for de_DE is `.`. Everything I saw on the web says that. I tested this when I made the change in my WSL. Having said that, there may be a way to hard code your separator.

> How can I write a test that ensures that the system locale is respected?

I believe these are just env var settings like you say with LC_ALL. So, maybe you can try to set an env var during testing?

x3rAx commented:
> Thanks for working on this!

It's my pleasure :)

> > Where is the best place to move that code so that it is not duplicated?
> 
> I'd put it in nu-utils as long as it has no nushell types like Value

Alright, I made a `nu_utils::get_system_locale()` which returns a `Locale` and made it so that it defaults to 'en_US'  when the system locale can not be found.

Would it be better to return an `Option` so the user of the function can decide which locale to fall back to? Or should I rename the function so that it is clear that it has a fallback? Or do you think it is sufficient as it now is?
 
> > How can I make sure the tests work independently of the system locale?
> > When the system locale is set to de_DE, the decimal separator is a , (comma). When running the tests with LC_ALL set to de_DE, the tests fail because they expect a . (period) as decimal separator.
> 
> I'm pretty sure the decimal separator for de_DE is `.`. Everything I saw on the web says that. I tested this when I made the change in my WSL. Having said that, there may be a way to hard code your separator.

I'm from Germany so I'm pretty sure it is `,`. :sweat_smile: The `.` is our thousands separator.

> > How can I write a test that ensures that the system locale is respected?
> 
> I believe these are just env var settings like you say with LC_ALL. So, maybe you can try to set an env var during testing?

Okay I figured it out: I already tried to set `LC_ALL` within the test but the tests kept failing from time to time, even when I restored the value of `LC_ALL` after the test. This led me to the conclusion that there must be some kind of race condition: The tests are executed in parallel. If I modify the environment of one test and another test is then executed in parallel, the environment of that second test is also modified.

I've now added a `with_fake_locale()` function that takes a locale string and a closure, modifies `LC_ALL`, executes the closure and restores the previous value of `LC_ALL`. It uses a `Mutex` to ensure that at most one closure is run at the same time while `LC_ALL` is modified.

fdncred commented:
> I'm from Germany so I'm pretty sure it is ,. 😅 The . is our thousands separator.

Sorry, I thought we were talking about thousands separator. I was confused.

I think the rest or your changes are fine. Good work! However, I'm worried about one thing. Is there a way to put the tests somewhere else? I don't like nu-test-support being included in nu-utils. It included nu-glob and nu-path. It may be ok, but I want to do everything possible to prevent having nu-* in nu-utils so that we can include nu-utils in every crate if we need to without recursive dependency issue.

x3rAx commented:
> > I'm from Germany so I'm pretty sure it is ,. sweat_smile The . is our thousands separator.
> 
> Sorry, I thought we were talking about thousands separator. I was confused.

Don't worry about it :smile_cat: 

> I think the rest or your changes are fine. Good work!

Thank you!

> However, I'm worried about one thing. Is there a way to put the tests somewhere else? I don't like nu-test-support being included in nu-utils. It included nu-glob and nu-path. It may be ok, but I want to do everything possible to prevent having nu-* in nu-utils so that we can include nu-utils in every crate if we need to without recursive dependency issue.

I see, where do you think the tests should go then? Or should we move the `with_fake_locale()` to nu-utils? But I can see it being used in other tests as well so I think nu-test-support would be a place where it is more easily found.

fdncred commented:
> I see, where do you think the tests should go then?

Can we just find a place in nu-command somewhere?

x3rAx commented:
> Can we just find a place in nu-command somewhere?

Sure I can move it there but since this is not testing a command but a `nu-utils` function, it feels a little bit out of place. How about putting it in `nu-test-support`? It's not exactly a test-support function either but one could argue that the test-support is there to support testing and therefore, it supports in testing this function without having to depend on `nu-*` in `nu-utils`. :sweat_smile: What do you think?

fdncred commented:
> How about putting it in nu-test-support?

I'm ok with that.

fdncred commented:
@x3rAx are we ready to land once this ci goes green?

x3rAx commented:
> @x3rAx are we ready to land once this ci goes green?

I'd say yes but I don't know why the test fails and I don't have access to a macOS to debug it. But what I find strange is that the tests `test_get_system_locale_de` and `test_get_system_locale_en` are not contained in the [raw macOS log](https://pipelines.actions.githubusercontent.com/serviceHosts/a4726ede-0a25-44e5-9ef0-eb645abb9d97/_apis/pipelines/1/runs/2782/signedlogcontent/7?urlExpires=2022-07-23T22%3A52%3A52.4371694Z&urlSigningMethod=HMACV1&urlSignature=7rNePfVhw%2BQfHPVPCZKEJ%2BccgCuxQCzX972E5uYRnVA%3D). The [raw Ubuntu logs](https://pipelines.actions.githubusercontent.com/serviceHosts/a4726ede-0a25-44e5-9ef0-eb645abb9d97/_apis/pipelines/1/runs/2783/signedlogcontent/6?urlExpires=2022-07-23T22%3A52%3A47.2181064Z&urlSigningMethod=HMACV1&urlSignature=vmZFSv%2B0%2BHkVplMZZqA5Tb8UKK8RPTWM8uiQY4OkWoU%3D) do contain these tests.

fdncred commented:
@x3rAx I tried to get this to work in macOS and couldn't. I'm not sure why.
This is in zsh.
```
fdncred@M1X nushell % locale
LANG="de_DE.UTF-8"
LC_COLLATE="de_DE.UTF-8"
LC_CTYPE="de_DE.UTF-8"
LC_MESSAGES="de_DE.UTF-8"
LC_MONETARY="de_DE.UTF-8"
LC_NUMERIC="de_DE.UTF-8"
LC_TIME="de_DE.UTF-8"
LC_ALL="de_DE.UTF-8"
```
I run the test and still get
```
thread 'commands::str_::into_string::int_into_string_decimals_respects_system_locale_de' panicked at 'assertion failed: `(left == right)`
  left: `"10.0"`,
 right: `"10,0"`', crates/nu-command/tests/commands/str_/into_string.rs:253:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test commands::str_::into_string::int_into_string_decimals_respects_system_locale_de ... FAILED
```

@jntrnr any ideas how to get this to work on Mac?

x3rAx commented:
@fdncred the tests *should* work without exporting `LC_ALL`. That's what the `with_fake_locale("de_DE.UTF-8", || { ... })` is for ([this line](https://github.com/x3rAx/nushell/blob/f6cf3d16b628948bc28d1ff3011c8f3d0fd92082/crates/nu-command/tests/commands/str_/into_string.rs#L245)). Maybe the `sys-locale` crate works differently on macOS... But then I'd expect the `test_get_system_locale_de` to fail as well.

Could you run

```
cargo test -p nu-test-support test_get_system_locale
```

and tell me if the two tests are green?

fdncred commented:
> Could you run

```
cargo test -p nu-test-support test_get_system_locale 
   Compiling num-complex v0.2.4
   Compiling num-bigint v0.2.6
   Compiling num-iter v0.1.43
   Compiling num-bigint v0.4.3
   Compiling num-rational v0.2.4
   Compiling num v0.2.1
   Compiling hamcrest2 v0.3.0
   Compiling nu-test-support v0.65.1 (/Users/fdncred/src/forks/nushell/crates/nu-test-support)
    Finished test [unoptimized + debuginfo] target(s) in 1.87s
     Running unittests src/lib.rs (target/debug/deps/nu_test_support-4b9dab4fbac04b6d)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 3 filtered out; finished in 0.00s
```

fdncred commented:
If you look at these implementations here, it appears that linux works with the env vars but Mac, which is bsd based, does not work with env vars and works differently https://github.com/bcmyers/num-format/tree/master/num-format/src/system_locale/nix

Here's the way google thinks is the proper way to get the macOS locale.
```
> defaults read -g AppleLanguages
(
    "en-US"
)
```
However, we shouldn't use the macOS `defaults` command to write different languages because it's a system level setting and I don't think Mac owners, myself included, want tests changing my personal locale each time we run tests. The real question is, when a Mac owner has their locale set to de_DE, does this change work.

x3rAx commented:
@fdncred Thank you!

Okay so the `sys-locale` crate does not use the `LC_*` environment variables directly on macOS. I Found [this line](https://github.com/1Password/sys-locale/blob/1b0e55908620e1ed10dc2585df47a5bc5d04cf55/src/apple_locale.m#L5) where it uses the OS library to get the first "preferredLanguage" (compare it to [this section](https://github.com/1Password/sys-locale/blob/1b0e55908620e1ed10dc2585df47a5bc5d04cf55/src/unix.rs#L30-L33) of the `unix.rs`). I don't know how the OS lib determines the language, but it seems to ignore `LC_ALL` (otherwise the test should have passed after you exported `LC_ALL`).

Another reason might be that the locale for `de_DE` is simply not installed. Could you check if

```
locale -a | grep de_DE
```

does output anything?

And could you also again run the tests (from bash) using:

```
locale=de_DE.UTF-8

export LANG=$locale LC_CTYPE=$locale LC_NUMERIC=$locale LC_TIME=$locale LC_COLLATE=$locale LC_MONETARY=$locale LC_MESSAGES=$locale LC_PAPER=$locale LC_NAME=$locale LC_ADDRESS=$locale LC_TELEPHONE=$locale LC_MEASUREMENT=$locale LC_IDENTIFICATION=$locale LC_ALL=$locale

locale

cargo test -p nu-test-support test_get_system_locale -- --test-threads=1
```

This exports `LANG` and all `LC_*` vars (except `LC_ALL`) as `de_DE.UTF-8`. If this does not affect the test results, I'd suggest disabling the tests using `with_fake_locale` on `target_os = "macos"` for now.

I won't have much time until about in the middle of October so I'd be glad to finalize this PR and create an issue to fix the tests on macOS later. I might also get access to a MacBook in October so maybe I can then better debug it.

x3rAx commented:
@fdncred Oh, you posted 2 minutes before me, I didn't see that.

> However, we shouldn't use the macOS defaults command to write different languages because it's a system level setting and I don't think Mac owners, myself included, want tests changing my personal locale each time we run tests.

I agree. I also didn't want my system settings to be changed by running tests, even if it would only be temporary ([Murphy's law](https://en.wikipedia.org/wiki/Murphy%27s_law)).

> The real question is, when a Mac owner has their locale set to de_DE, does this change work.

That's what my 2nd question from above (https://github.com/nushell/nushell/pull/6085#issuecomment-1192652399) was about. "How can I make sure the tests work independently of the system locale?"

I see the following solutions:

1. In the `get_system_locale()` function (which wraps the `sys-locale` crate), we could check if `LC_ALL` (or another env var like `TEST_LOCALE`) is set and return that locale. Doing so, we can use the `with_fake_locale()` test helper on any system for any locale. It would also allow a developer to try out nushell with different locales
2. Like the above, but using having two implementations of `get_system_locale()` one for `#[cfg(test)]` and one for `#[cfg(not(test))]` where the `test` implementation uses `LC_ALL` or defaults to `en_US.UTF-8` and the `not(test)` implementation uses the `sys-locale` crate.
3. Maybe even better: Wrap the `sys_locale::get_locale()` call in a simple (private) function that during `test` returns the value of the env var and during `not(test)` returns the result of `sys_locale::get_locale()`. Doing this, the logic of `get_system_locale()` can still be tested and the wrapper function is small enough to simply see that it would work without needing an extra test.
4. We could also skip the tests on macOS. (I don't like this as a permanent solution)
5. Maybe the `sys_locale::get_locale()` can be mocked but I have no experience with mocking in Rust and as I said, until mid October, I don't have time for digging into that.

I think I like variant 3 the most.

x3rAx commented:
We could also define a global Mutex during `test` instead of an env var. `with_fake_locale()` already works with a Mutex so we could simply make it public during testing and use it's value in the above mentioned wrapper function.

fdncred commented:
answers
```
> locale -a | grep de_DE

de_DE.UTF-8
de_DE.ISO8859-15
de_DE.ISO8859-1
de_DE
```
```
> export LANG=$locale LC_CTYPE=$locale LC_NUMERIC=$locale LC_TIME=$locale LC_COLLATE=$locale LC_MONETARY=$locale LC_MESSAGES=$locale LC_PAPER=$locale LC_NAME=$locale LC_ADDRESS=$locale LC_TELEPHONE=$locale LC_MEASUREMENT=$locale LC_IDENTIFICATION=$locale LC_ALL=$locale

> locale

LANG="de_DE.UTF-8"
LC_COLLATE="de_DE.UTF-8"
LC_CTYPE="de_DE.UTF-8"
LC_MESSAGES="de_DE.UTF-8"
LC_MONETARY="de_DE.UTF-8"
LC_NUMERIC="de_DE.UTF-8"
LC_TIME="de_DE.UTF-8"
LC_ALL="de_DE.UTF-8"
```
```
cargo test -p nu-test-support test_get_system_locale -- --test-threads=1
   Compiling nu-path v0.65.1 (/Users/fdncred/src/forks/nushell/crates/nu-path)
   Compiling nu-utils v0.65.1 (/Users/fdncred/src/forks/nushell/crates/nu-utils)
   Compiling nu-test-support v0.65.1 (/Users/fdncred/src/forks/nushell/crates/nu-test-support)
    Finished test [unoptimized + debuginfo] target(s) in 0.94s
     Running unittests src/lib.rs (target/debug/deps/nu_test_support-7775f594f6cedd30)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 3 filtered out; finished in 0.00s

     Running tests/get_system_locale.rs (target/debug/deps/get_system_locale-b7b288f66d908fcc)

running 2 tests
test test_get_system_locale_de ... FAILED
test test_get_system_locale_en ... ok

failures:

---- test_get_system_locale_de stdout ----
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"en"`,
 right: `"de"`', crates/nu-test-support/tests/get_system_locale.rs:20:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    test_get_system_locale_de

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

error: test failed, to rerun pass '-p nu-test-support --test get_system_locale'
```

x3rAx commented:
@fdncred Thank you for your support.

I'll implement option 3 in the next few days.

fdncred commented:
@x3rAx still working on this?

x3rAx commented:
@fdncred Yes, I am. I just didn't have much time and had to find a slightly different solution than the one I suggested.


Firstly because the `#[cfg(test)]` and `#[cfg(not(test))]` didn't work as I imagined, secondly because I didn't know that the `nu!` macro actually runs the `nu` binary. When I saw that, I wanted to change the macro so that you can specify `locale` as an option. That was not so easy and it turned out that I had to essentially rebuild it but on the positive side I know **a lot more** about macros now :D

Since the `nu!` macro calls the binary directly, there is no need to change the environment variable of the current process. So I adapted the macro, so that now on the one hand `pwd` is optional (default value is now `"."`) and on the other hand the `locale` can be set. I added comments to the macro pattern matchers, which hopefully explain understandable enough what happens.

You can now use it like this for example:

```nu
nu!(
    locale: "en_DE.UTF-8",
    "10 | into string --decimals 2"
)
```

I think it would also make sense to set the locale to "en_US.UTF-8" by default if it is not set otherwise? This should also make the tests run more reliably on a non-en_US system.

x3rAx commented:
I'll do the rebase / merge conflicts tomorrow

fdncred commented:
Wow! Good work. Agreed to "en_US.UTF-8" by default.

x3rAx commented:
@fdncred I think that's it now. It should now also work on macOS

@nibon7 I've noticed that you used the `nu!` macro without passing `cwd` as `cwd: cwd`. When this PR gets merged, this won't work anymore because `cwd` is then optional (defaulting to "."). I actually searched the code and there was no instance of a `nu!` macro call without the `cwd:` so I figured it was ok to remove it. I hope you are fine with it

fdncred commented:
@x3rAx I see it works for me in en_US but I can't get it to work on my Mac with de_DE.
```
> export LANG="de_DE.UTF-8"
> locale
LANG="de_DE.UTF-8"
LC_COLLATE="de_DE.UTF-8"
LC_CTYPE="de_DE.UTF-8"
LC_MESSAGES="de_DE.UTF-8"
LC_MONETARY="de_DE.UTF-8"
LC_NUMERIC="de_DE.UTF-8"
LC_TIME="de_DE.UTF-8"
LC_ALL="de_DE.UTF-8"

> 10 | into string -d 1
10.0 <-- shouldn't this be 10,0?
```
I then run
```
> cargo test -p nu-test-support test_get_system_locale
...
running 2 tests
test test_get_system_locale_de ... ok
test test_get_system_locale_en ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

x3rAx commented:
> ```
> > export LANG="de_DE.UTF-8"
> ```

Mac has a different system for getting the locale, remember my comment from above:

> Okay so the sys-locale crate does not use the LC_* environment variables directly on macOS. I Found [this line](https://github.com/1Password/sys-locale/blob/1b0e55908620e1ed10dc2585df47a5bc5d04cf55/src/apple_locale.m#L5) where it uses the OS library to get the first "preferredLanguage" (compare it to [this section](https://github.com/1Password/sys-locale/blob/1b0e55908620e1ed10dc2585df47a5bc5d04cf55/src/unix.rs#L30-L33) of the unix.rs).

There are definitively cli tools that respect `LC_*` (for example `locale` is one of them ^^) but maybe that's because they originated in the Linux world and were brought to macOS through homebrew or something (I honestly have no idea, just wild guessing). But it seems like the official way is to ask the OS for the users "preferred language".

In a non-release build you should be able to override the locale using the `NU_TEST_LOCALE_OVERRIDE` env var:

```
> export LC_ALL="en_US.UTF-8"
> locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8
>
> # So all LC_ vars are now en_US
> # Now export override to de_DE (which uses comma as decimal separator)
>
> export NU_TEST_LOCALE_OVERRIDE="de_DE.UTF-8"
> cargo run
〉10 | into string -d 1
10,0
```

fdncred commented:
ok, let's try this. thanks!

sholderbach commented:
No worries @x3rAx 
Thanks for the thorough inverstigations you have already done into that topic! 
Yeah we have to look into the general issue of localization more thoroughly. 
The patch sees like a good enough work around if it bugs us now.
The `Example` testing might need a broader overhaul, so that's a good point to keep in mind.
